pkg/solver/server.go

```diff
@@ -62,17 +63,44 @@ func NewSolverServer(
  *
 */
 
+func exemptIPKeyFunc(exemptIPs []string) func(r *corehttp.Request) (string, error) {
+	return func(r *corehttp.Request) (string, error) {
+		ip, _, err := net.SplitHostPort(r.RemoteAddr)
+		if err != nil {
+			ip = r.RemoteAddr
```

Comment:
Using `r.RemoteAddr` won't work because we proxy through a Cloudflare tunnel. We need to determine the real IP address or we end up rate limiting based on the proxy IP address.

The `httprate` implementation of `KeyByRealIP` handles this correctly: https://github.com/go-chi/httprate/blob/9d627fb3c8a009b20ae198abe84b3a1d10a22a94/httprate.go#L52-L74. (Cloudflare adds an extra header we can key on.)

There's also a `canonicalizeIP` call in there. Haven't looked at it, but there may be things we want in it.